### PR TITLE
6 more naive methods for Tensor Primitives.

### PR DESCRIPTION
**Pull request by michaelgsharp**
Added in naive implementations for
* CosineSimilarty
* Distance
* Dot
* Normalize
* SoftMax
* Sigmoid

@tannergooding @stephentoub @jeffhandley should we rename normalize as discussed since its not really the operation we are doing?

### CODE REVIEW

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (gfoidl) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:270]: Don't use `var` in runtime-repo. Use the concrete type here instead.
Same on other places.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (gfoidl) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:278]:
```suggestion
                magx += x[i] * x[i];
                magy += y[i] * y[i];
```
? No need to call pow.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (gfoidl) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:301]:
```suggestion
                float dist = x[i] - y[i];
                distance += dist * dist
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (gfoidl) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:341]:
```suggestion
                magx += x[i] * x[i];
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:4]: What is this needed for?

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:366]: This doesn't look right. Did you intend for this to be x.Length? Assuming it's wrong, that also suggests a test hole.

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:718]: These tests will be insufficient for vectorized implementations as they're not large enough to trigger paths for some vector widths. It'd also be good to validate empty inputs.

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:790]: It'd be good to add tests where the destination is longer the input. I expect if you do that, you'll find it fails with the current implementation.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:262]: We should document the ArgumentException that might be thrown (here and in the other methods).

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:335]: L2Normalize? NormalizeL2?

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:4]: Shouldn't be needed, was probably added automatically when I did a quick fix or something. Removed.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:327]: This will become a one-liner on top of the Aggregate helper in https://github.com/dotnet/runtime/pull/92154

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:262]: Added the exception to all the docs.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:304]: This will become a one-liner on top of the Aggregate helper in https://github.com/dotnet/runtime/pull/92154

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:270]: Fixed.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:278]: Fixed

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:301]: Fixed

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:335]: Changed it to L2Normalize.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:341]: Fixed.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:366]: Yeah, I meant it to be x and it was a whole, and I don't have any tests where the dest is longer than the input. Let me add some.

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:718]: Yeah, I know they will be insufficient for the vectorized tests, but since none of these are yet I wasn't too worried. Would you like me to fix that now though or when we get the vectorized implementations in? Happy either way.

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:790]: Yeah, it would have failed. Added in those tests and fixed it.

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs
Comment (stephentoub) [reply src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:718]: Let's get it merged and then we can add more.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:356]: Would you like me to wait till that PR is merged to make that change? or just do it after? Same question for the comment above.

Comment (stephentoub) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:356]: I'll include it in that PR once this merges.

**Review by ericstj — APPROVED** [review]
LGTM

**Review by stephentoub — APPROVED** [review]
I'm going to merge this and then fix up a few things after rebasing my PR on it.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (jeffhandley) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:285]: If all values are 0 in one of the spans, I think that'd result in a DivideByZero here. Is that the desired behavior?

**Review by jeffhandley — COMMENTED** [review]
Echoing the existing feedback about tests, we could add a lot more tests for all these methods to capture edge/corner cases and more exceptional data too.

_I meant to say, but we can add more tests after the RC2 milestone/snap_

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (tannergooding) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:285]: Floating-point doesn't produce `DivideByZeroException`, it returns `NaN` (or `+/-Infinity`, depending on the exact values involved).

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:269]: We only need one of the checks here since we've already determined the lengths are equal.

That being said, is it desirable to throw here rather than allowing it to just return `NaN`?

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:302]: Same here, the second check is redundant since we know the lengths are equivalent.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:347]: Why is this `L2Normalize`? The standard API computes the "euclidean norm", that is the length or magnitude of the vector.

The name should likely just be `Norm`

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:302]: These were already changed in my pr.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:381]: Why `MathF.Pow` and not `MathF.Exp`?

Why `(float)Math.E` and not `MathF.E`?

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (jeffhandley) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:285]: Never mind. Today I learned, "Dividing a floating-point value by zero doesn't throw an exception; it results in positive infinity, negative infinity, or not a number (NaN), according to the rules of IEEE 754 arithmetic."

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:381]: @tannergooding, can you submit a cleanup PR with the changes you think should be made? Thanks.

**Pull request merged**
